shor/gates.py

```diff
@@ -504,6 +504,35 @@ def to_matrix(self) -> np.ndarray:
         return np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, np.exp(2*1j*np.pi/2**self.k)]], dtype='complex')
 
 
+class qModExp(_Gate):
+    # map input registers to a set of output registers, where f(x) = a^x mod N
+    symbol = 'qModExp'
+    def __init__(self, a, N, *qubits, **kwargs):
+        self.a = a
+        self.N = N
+        kwargs['dimension'] = 1
+        if not qubits:
+            qubits = [0]
+
+        super().__init__(*qubits, **kwargs)
+
```

Comment:
There is also a to_gates method. Intention here is to convert a layer / sub circuit into base gates.

P.S. we were trying to get this working with just 15 earlier here:
https://github.com/shor-team/shor/blob/master/shor/algorithms/shor.py#L70

But yes, the tests aren't currently passing, grateful for the help here!